File: src/Functional/Erlang/composite.erl
-module(composite).
-export([main/0]).

size({file, Bytes}) -> Bytes;
size({folder, Children}) -> lists:sum([size(Child) || Child <- Children]).

main() ->
    Readme = {file, 2},
    Docs = {folder, [{file, 3}, {file, 5}]},
    Root = {folder, [Readme, Docs]},
    io:format("leaf=~p~n", [size(Readme)]),
    io:format("docs=~p~n", [size(Docs)]),
    io:format("root=~p~n", [size(Root)]).

File: src/Functional/Elixir/composite.exs
defmodule Composite do
  def file(bytes), do: {:file, bytes}
  def folder(children), do: {:folder, children}

  def size({:file, bytes}), do: bytes
  def size({:folder, children}), do: Enum.reduce(children, 0, fn child, total -> total + size(child) end)
end

readme = Composite.file(2)
docs = Composite.folder([Composite.file(3), Composite.file(5)])
root = Composite.folder([readme, docs])

IO.puts("leaf=#{Composite.size(readme)}")
IO.puts("docs=#{Composite.size(docs)}")
IO.puts("root=#{Composite.size(root)}")
